File: Solutions/Python/copy_list_with_random_pointer.py
"""
    [MEDIUM]
    138. Copy List with Random Pointer

    Concepts:
    - linked list
    - hash table

    Stats:
        Runtime | 36 ms     [Beats 70.78%]
        Memory  | 17.41 MB  [Beats 62.71%]
"""

"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None

        orig_to_cpy = {} # Node_Object -> node.val

        # associate each node object with a
        # new node of same value
        curr = head
        while curr:
            orig_to_cpy[curr] = Node(curr.val)
            curr = curr.next

        # link all new nodes together using
        # `next` and `random` of old nodes
        curr = head
        while curr:
            # .get() : returns value with given key or None if not in dict
            orig_to_cpy[curr].next = orig_to_cpy.get(curr.next)
            orig_to_cpy[curr].random = orig_to_cpy.get(curr.random)
            curr = curr.next

        # return copy of new list
        return orig_to_cpy[head]

File: Solutions/Cpp/copy_list_with_random_pointer.cpp
/*
    [MEDIUM]
    138. Copy List with Random Pointer

    Concepts:
    - linked list
    - hash table

    Stats:
        Runtime | 9 ms      [Beats 21.24%]
        Memory  | 15.15 MB  [Beats 50.27%]
*/

/*
// Definition for a Node.
class Node {
public:
int val;
Node* next;
Node* random;

Node(int _val) {
    val = _val;
    next = NULL;
    random = NULL;
}
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (head == nullptr)
            return nullptr;

        // Node_Object -> node.val
        std::unordered_map<Node*, Node*> orig_to_cpy = {};
        Node* curr;

        // associate each node object with a
        // new node of same value
        curr = head;
        while (curr != nullptr) {
            orig_to_cpy[curr] = new Node(curr->val);
            curr = curr->next;
        }

        // link all new nodes together using
        // `next` and `random` of old nodes
        curr = head;
        while (curr != nullptr) {
            orig_to_cpy[curr]->next = (orig_to_cpy.contains(curr->next)) ?
            (orig_to_cpy[curr->next]) : (nullptr);
            orig_to_cpy[curr]->random = (orig_to_cpy.contains(curr->random)) ?
            (orig_to_cpy[curr->random]) : (nullptr);

            curr = curr->next;
        }

        // return cop of new list
        return orig_to_cpy[head];
    }
};